[20.09] firefox: enable pipewire+webrtc for wayland users

worldofpeace opened the pull request:
(cherry picked from commit 31e54cdde34073b3b83e9fca4434cb0cedcb2829)

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Backports https://github.com/NixOS/nixpkgs/pull/84233

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

colemickens committed:
firefox: enable pipewire+webrtc for wayland users

(cherry picked from commit 31e54cdde34073b3b83e9fca4434cb0cedcb2829)

endgame commented on pkgs/applications/networking/browsers/firefox/common.nix:
```diff
@@ -94,6 +94,11 @@ stdenv.mkDerivation ({
 
   patches = [
     ./env_var_for_system_dir.patch
+    (fetchpatch {
```
Should this only be in `patches` if `waylandSupport`?

andir replied:
We should have less conditional patches as that makes it harder to have silent breakage. Usually those features should only be disabled via the build system and not via conditional patches.

andir replied:
After reading the patch, which doesn't support making pipewire optional, I am actually in favor of just dropping it or making it at least non-default. I am even less motivated to maintain this now then I was before looking at the patch. It seems like it needs to at least contain one more feature flag check in `media/webrtc/trunk/webrtc/modules/desktop_capture/desktop_capture_generic_gn/moz.build`. Probably the same that was applied in `media/webrtc/trunk/webrtc/modules/desktop_capture/BUILD.gn`. And `pipewire` and `spa` should be discovered via pkg_config (like in the other case) not just via hardcoded paths.

endgame commented on pkgs/applications/networking/browsers/firefox/common.nix:
```diff
@@ -137,6 +142,11 @@ stdenv.mkDerivation ({
     rm -rf obj-x86_64-pc-linux-gnu
+
+    # needed for enabling webrtc+pipewire
+    substituteInPlace \
```
Should this only run if `waylandSupport`?

andir replied:
That is what I added.

endgame commented on pkgs/applications/networking/browsers/firefox/common.nix:
```diff
@@ -137,6 +142,11 @@ stdenv.mkDerivation ({
+    # needed for enabling webrtc+pipewire
+    substituteInPlace \
+      media/webrtc/trunk/webrtc/modules/desktop_capture/desktop_capture_generic_gn/moz.build \
+      --replace /usr/include ${pipewire.dev}/include
```
Do other libraries need their include dirs listed here?

andir replied:
Why would they?

andir replied:
Looking at the patch we should probably also patch at least one other library:
```
+    CXXFLAGS += [ "-I/usr/include/pipewire-0.3" ]
+    CXXFLAGS += [ "-I/usr/include/spa-0.2" ]
```

andir replied:
Ok, these two both come from `pipewire` to this patching is actually correct.

endgame reviewed (commented):
I don't know the firefox build system so some of these comments might be off-base.

worldofpeace commented:
@endgame This is a backport. That means you really can't request changes to already committed code without encouraging diversion which has to be forward ported.

endgame commented:
@worldofpeace Fair enough then.

worldofpeace commented:
@endgame Not sure if you could submit a  review the original merged PR, though. Let's see what @andir thinks

endgame commented:
I tried to build firefox from nixpkgs master without wayland and it broke because it couldn't find `-lpipewire` so  something needs to be reported: either on the original PR or as a new issue. No point pretending that wayland is an option if it breaks when turned off.

andir commented:
Thank you for testing and reporting @endgame.

@worldofpeace due to the quality of the patch I didn't want it in the first place but was probably too inactive to make my opinion voiced enough. At this stage I'd at least refrain from adding it to the stable release. It is a nice feature and a pity but I do not see us having the manpower to do fix and support this patch.

I'll spend a bit of time trying to clean the patch up to fix the situation in both master and this backport. If that fails I'm still in favor of skipping it or making the entire code optional (which is a mess I really dislike).

worldofpeace commented:
This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium

andir committed:
firefox: ensure pipewire is only enabled with wayland & webrtc

This ensures that we aren't applying any of the experiemental pipewire
patches when the dependencies aren't enabled. As of now pipewire only
works with wayland and webrtc. If either of them are not activated we
can't build with pipewireSupport and we should not.

andir commented:
> This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium

Yeah but most of them hide it behind a feature flag and/or only provide a single build for their entire distribution. Here we are trying to allow users to customize the package but didn't bake that into the derivation when this patch was added.

I just pushed c9c0b90b62bb863b56a7bdc356d38f7b143f2fa5 to this branch which does the required housekeeping for the added patch. I'll also add this to unstable once we agree on it.

worldofpeace commented:
> > This patch is actually being used in more distros than I thought, see https://github.com/emersion/xdg-desktop-portal-wlr/wiki/Screencast-Compatibility#webrtc-aka-firefoxchromium
> 
> Yeah but most of them hide it behind a feature flag and/or only provide a single build for their entire distribution. Here we are trying to allow users to customize the package but didn't bake that into the expression when this patch was added.
> 
> I just pushed [c9c0b90](https://github.com/NixOS/nixpkgs/commit/c9c0b90b62bb863b56a7bdc356d38f7b143f2fa5) to this branch which does the required housekeeping for the added patch. I'll also add this to unstable once we agree on it.

Thanks, it looks good.

worldofpeace commented:
@andir Could you also summarize your current stance on this? It seems you've taken a few and it's not clearly linear because some of it is in the review comments.

andir commented:
This PR as is is probably fine. If you want to carry that into a release that sounds fine. I still worry about keeping the feature stable as we effectively now depend on Mozilla and a single fedora maintainer. Given the history with pipewire fedora will probably continue to provide patches. I just hope they'll do it on time as very often they were slower than us in updating the browser when security issues appeared.
